articles/python-generators/index.md

```diff
@@ -0,0 +1,160 @@
+
+**Generator Object** uses lazy evaluations to yield sequences. They are used either by calling the next method on the  generator object or using the generator object in a “for in” loop.
+
+**Yield** statement suspends function’s execution and sends a value back to the caller, but retains enough state to enable function to resume  where it is left off. When resumed, the function continues execution  immediately after the last yield run. This allows its code to produce a  series of values over time, rather than computing them at once and  sending them back like a list.
+
+**Generator Expressions** are a high-performance, memory–efficient generalization of list comprehensions and generators.
+
+### Generator Expressions vs List Comprehensions
+```python
+list_comprehension = ['Hello world' for i in range(3)]
+generator_expression = ('Hello world' for i in range(3))
+```
+
+Unlike list comprehensions, generator expressions don’t  construct list objects. Instead, they generate values “just in time”  like a class-based iterator or generator function would.
+
+To access the values produced by the generator expression, you need to call `next()` on it, just like you would with any other iterator. When there isn't next value in the generator object, It throws a `StopIteration` exception. 
+
+```python
+>>> next(generator_expression)
+'Hello world'
+>>> next(generator_expression)
+'Hello world'
+>>> next(generator_expression)
+'Hello world'
+>>> next(generator_expression)
+StopIteration
+```
+A for loop can be used to iterate the generator object.
+
+```python
+for expression in generator_expression:
+    print(expression)
+```
+
+
+
+Given a range of n numbers to filter out odd numbers. We could solve the problem using three different methods.
+1. Using function.
+
+```python
+def even_integers_function(n):
+    result = []
+    for i in range(n):
+        if i%2==0:
+            result.append(i)
+    return result
+```
+
+```python
+>>>even_integers_function(10)
+[0,2,4,6,8,10]
+```
+
+2. Using Generator function. 
+   The generator function below yields a generator object which we can iterate through to get the list of even numbers. 
+
+```python
+def even_integer_generator(n):
+    for i in range(n):
+        if i % 2 == 0:
+            yield i
+```
+A generator function returns a generator object which can be iterated to get the values.
+```python
+>>>even_integer_generator(10)
+<generator object even_integer_generator at 0x1036c3200>
+```
+
+```python
+>>>list(even_integer_generator(10))
+[0,2,4,6,8,10]
+```
+3. Using Generator expression.
+```python
+even_integers = (i for i in range(n) if i%2==0)
+```
+Given a list of names to convert each name in the list to uppercase.
```

Comment:
Rephrase the sentence to-> Given a list of names, we can convert each name in the list to uppercase using the following code: